site/en/docs/android/custom-tabs/guide-warmup-prefetch/index.md

```diff
@@ -0,0 +1,106 @@
+
+```xml
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools">
+  …
+   <queries>
+        <intent>
+            <action android:name="android.support.customtabs.action.CustomTabsService" />
+        </intent>
+    </queries>
+</manifest>
+```
+
+Here is the full code:
+
+```java
+private CustomTabsClient mClient;
+private CustomTabsSession mSession;
+
+private CustomTabsServiceConnection mConnection = new CustomTabsServiceConnection() {
+    @Override
+    public void onCustomTabsServiceConnected(
+            @NonNull ComponentName name,
+            @NonNull CustomTabsClient client
+    ) {
+        mClient = client;
+        // Warm up the browser process
+        mClient.warmup(0 /* placeholder for future use */);
+        // Create a new browser session
+        mSession = mClient.newSession(new CustomTabsCallback());
+        // Optional: pre-render pages the user is likely to visit
+        // you can do this any time while the service is connected
+        mSession.mayLaunchUrl("https://developers.android.com", null, null);
+    }
+
+    @Override
+    public void onServiceDisconnected(ComponentName name) {
+        mClient = null;
+        mSesssion = null;
+    }
+};
+
+private void bindCustomTabService(Context context) {
+    // Check for an existing connection
+    if (mClient != null) {
+        // Do nothing if there is an existing service connection
+        return;
+    }
+
+    // Get the default browser package name, this will be null if
+    // the default browser does not provide a CustomTabsService
+    String packageName = CustomTabsClient.getPackageName(context, null);
+    if (packageName == null) {
+        // Do nothing as service connection is not supported
+        return;
+    }
+    CustomTabsClient.bindCustomTabsService(context, packageName, mConnection);
+}
+
+@Override
+protected void onCreate(Bundle savedInstanceState) {
+…
+    bindCustomTabService(this);
+    findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
+        @Override
+        public void onClick(View view) {
+            String url = "https://developers.android.com";
+            CustomTabsIntent intent = new CustomTabsIntent.Builder(session)
+                    .build();
+            intent.launchUrl(MainActivity.this, Uri.parse(url));
+        }
+    });
+}
+```
+
+{% Aside "caution" %}
+A Custom Tabs service connection might fail while your activity is running. If your app requires an active service connection, there are two strategies to ensure a working service connection:
+
+1. When eagerly connecting to the CustomTabsService during `onCreate`, reconnect to the service  if your activity gets disconnected and the `onServiceDisconnected` callback gets invoked.
+2. Wait until the user hits a button, establish the service connection and then launch the custom tab.
```

Comment:
> Wait until the user hits a button

do you mean triggers an action to launch a CustomTab? I am not sure I understand what button is being hit

Reply:
rephrased